### PR TITLE
fix(api): preserve source files when deleting duplicate markers

### LINKED ISSUE #3627: [Bug]: dup- doc_status records reuse the original document's file_path, so file_path is not a unique document key
### Do you need to file an issue?

- [x] I have searched the existing issues and this bug is not already filed.
- [x] I believe this is a legitimate bug, not just a question or feature request.

### Describe the bug

Since #2591, uploading content that already exists writes a trackable FAILED record (`compute_mdhash_id(f"{doc_id}-{track_id}", prefix="dup-")`, `lightrag.py` ~1475) — with `status: FAILED`, `chunks_count: 0`, an `error_msg` naming the original doc id, and **the same `file_path` as the live document**.

The trackability itself is a good design (this issue is not asking to revert it). The problem is the unstated consequence: `file_path` stops being a unique key in `doc_status`. One file name can now map to a `doc-…` record (the live document) plus one or more `dup-…` markers.

### Impact

Any cleanup or bookkeeping keyed on a status record's `file_path` over-selects. Measured concretely: ingesting identical content twice under one name produced `doc-889c30…` (processed) plus `dup-c67cfb…` (failed); deleting the *marker* with name-keyed cleanup destroyed the live document's job/tracking records while the document itself remained — a confusing half-deleted state with no error.

### Expected Behavior

One or more of:

- document the invariant loss prominently (file_path is a display name, not a key);
- add a first-class discriminator field (e.g. `is_duplicate: true` at the top level rather than only inside `metadata`) so consumers can filter cheaply;
- exclude dup- markers from APIs that enumerate documents by file path.

We work around it by gating on `metadata.is_duplicate` (OR the `dup-` id prefix) before any name-keyed step.

### PR DESCRIPTION
**Pull request by pengpengyi92**
## Summary

- preserve the shared source file when deleting a duplicate-attempt `doc_status` marker with `delete_file=true`
- keep ordinary primary-document file deletion unchanged
- document that `DocProcessingStatus.file_path` is not unique and does not imply physical file ownership
- add focused regression coverage for both duplicate-marker and primary-document deletion

Closes #3627.

## Root cause

Duplicate-attempt rows intentionally reuse the primary document's canonical `file_path`. The API background deletion path called `adelete_by_doc_id()` first, which removed the status row, and then treated the returned `file_path` as proof that the deleted row owned the physical source file. Deleting a `dup-*` marker with `delete_file=true` could therefore remove the primary document's shared source file.

## Fix

When physical deletion is requested, the route reads the status metadata before `adelete_by_doc_id()` removes the row. It uses the existing `metadata.is_duplicate` discriminator to suppress only physical source-file cleanup for duplicate markers. The duplicate status row and its graph/cache data are still deleted through the existing path.

This deliberately avoids:

- changing the `DeletionResult` public model
- changing storage schemas or duplicate creation
- inferring duplicate status from an ID prefix
- weakening file deletion for ordinary documents

## Validation

Focused regression and deletion-contract tests:

```text
5 passed
```

Full document-route test file on Windows:

```text
82 passed, 3 failed
```

The same three failures reproduce unchanged on a clean `origin/main` worktree. They are existing Windows-only assumptions around symlink privileges and POSIX `0600` mode bits, unrelated to this change.

Additional checks:

- `ruff format --check` on all three changed files: passed
- Ruff findings restricted to newly added lines: none
- `git diff --check`: passed

### CODE REVIEW

**Comment by danielaskdd** [comment]
Reviewed the change against the three code paths that write
`metadata.is_duplicate`. The discriminator choice is the right one (the `dup-`
prefix would have been incomplete), but the predicate is broader than the
condition it is meant to express, and that introduces a new regression.

## 1. `metadata.is_duplicate` marks three populations, only two of which lack file ownership

The guard suppresses physical cleanup for every row carrying `is_duplicate`, but
that flag is written by three different producers with different file-ownership
semantics:

| Producer | Shares a live row's `file_path`? | Owns physical files? | Correct behavior |
|---|---|---|---|
| Enqueue `dup-*` markers (`pipeline.py:1395`) | yes | no | suppress ✔ |
| Source-conflict repair losers (`source_conflict_repair.py`) | yes, by definition | no, shared with the winner | suppress ✔ |
| Post-parse `content_hash` demotion (`pipeline.py:6166`) | **no** | **yes** | **should still delete ✘** |

The third case is the document's own `doc-*` row, flipped to FAILED in place.
Content-hash dedup fires only for *identical content under a different
filename* — a same-name collision is caught earlier by the filename rule — so
that row's `file_path` is unique to it. And `pipeline.py:6186` archives its
source into `__parsed__` immediately after the `is_duplicate` upsert, so the row
is the sole owner of that archived file plus the sidecar produced by the parse
that had already completed.

Verified on two worktrees with the same test (a demoted `doc-*` row whose
`file_path` is `twin.pdf`, its archived source in `__parsed__`, and an unrelated
primary `primary.pdf`):

```text
base (6f50fdd): 1 passed   # archived source + sidecar cleaned up
PR   (6968905): 1 failed
  left on disk: ['__parsed__/twin.pdf',
                 '__parsed__/twin.pdf.parsed/twin.blocks.jsonl',
                 'primary.pdf']
  INFO: Source file preserved for duplicate marker doc-twin: twin.pdf
```

The log line is also inaccurate here: `twin.pdf` belongs to no primary, only to
the row that was just deleted.

The consequence is worse than an ordinary leak. Once the delete completes the
status row is gone, so no API handle to that file remains — it is permanently
orphaned and can only be removed from the filesystem by hand. That also
contradicts `_file_path_for_parsed_artifact_dir`'s documented contract
(`document_routes.py:2183-2186`), which states raw bundles "are cleaned only when
the user deletes the document with `delete_file=True`"; this change removes that
sole exit.

This is still less severe than the bug being fixed (a leak rather than
destruction of live data), so the direction holds — but the predicate is worth
tightening before merge.

## 2. Suggestion: key on "does another row still claim this basename?"

The condition actually being expressed is not "is this row a duplicate" but
"after this row is gone, is anything else still referencing this file". Every
backend already implements a primary-only lookup for exactly this
(`base.py:1305`):

```python
if delete_file:
    other = await rag.doc_status.get_doc_by_file_basename(result.file_path)
    if other is not None and other[0] != doc_id:
        delete_physical_file = False  # a live document still uses this name
```

All three populations then fall out correctly: `dup-*` rows and conflict losers
find the surviving primary and suppress; the post-parse demotion — excluded from
the primary set by its own `is_duplicate` flag, with no same-name primary
anywhere — finds `None` and deletes normally. It also covers any future
name-sharing producer without another patch here.

Running it *after* `adelete_by_doc_id` is slightly cleaner (the row is gone, so
no self-comparison is needed) and means the extra read is only paid when
`delete_file=True` and the deletion actually succeeded.

## 3. The pre-read adds a per-document failure mode

`await rag.doc_status.get_by_id(doc_id)` sits inside the per-document `try`, so a
transient backend read error now sends the document to `failed_deletions`
*before* `adelete_by_doc_id` runs — a failure point that did not previously
exist. Failing closed is the right direction, but it is worth a comment, and a
batch delete with `delete_file=True` now pays N extra `doc_status` round-trips.
Moving the check after the deletion (see above) narrows both.

## 4. Minor

- **elif chain readability.** I enumerated all four combinations of
  (`delete_file`, is-duplicate, `file_path` valid) and the behavior is
  unchanged, so this is correctness-neutral. But `elif delete_physical_file:` is
  reachable only when `delete_physical_file` is true *and* `file_path` is
  missing, and mixing `delete_file` with `delete_physical_file` in one chain is
  easy to misread later.
- **`DeletionResult.file_path` still returns the shared path** for a duplicate
  row (`base.py:1721`, currently undocumented). Not changing the public model is
  reasonable, but the trap stays in place for the next caller; a one-line field
  comment mirroring the new `DocProcessingStatus.file_path` wording would close
  it cheaply.
- **Test gap.** The two new cases pin the two directions considered, but the
  post-parse demotion population is untested — that case doubles as the
  boundary documentation for this fix. (The `getattr(status_row, "metadata",
  None)` dataclass fallback is also untested, though per `get_by_id`'s
  `dict | None` contract it is unreachable and harmless.)
- **Docs.** `docs/FileProcessingPipeline.md` §7.3 tells operators that an
  explicit delete is the only way to remove these records; now that deletion
  behaves differently for source files depending on the row, that section could
  use a line.

For reference, `tests/api/routes/test_document_routes_docx_archive.py` is green
on Linux for both revisions (83 passed on base, 85 on this branch), which
matches the note that the three Windows failures are pre-existing and unrelated.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by pengpengyi92** [comment]
Addressed the review in 8da85c5.

- Replaced the broad `metadata.is_duplicate` predicate with a post-deletion ownership lookup via `get_doc_by_file_basename`.
- A `dup-*` row or source-conflict loser now preserves the source only when another primary still references that basename.
- A post-parse `doc-*` content duplicate with a unique basename deletes its own archived source and parsed sidecars.
- Moved the ownership read after successful status deletion and fail closed on lookup errors.
- Documented the shared nature of `DeletionResult.file_path` and the operator-visible deletion semantics.
- Added the missing post-parse demotion regression test and updated the duplicate-marker fixture to include its surviving primary.

Verification:
- Focused ownership tests: 3 passed.
- Route test file excluding the three Windows-only `upload_file_opener` permission/symlink cases: 83 passed, 3 deselected.
- The excluded cases require POSIX symlink/mode behavior and fail locally on Windows before reaching this change.
- `git diff --check` passes.

**Comment by danielaskdd** [comment]
@codex review

**Comment by danielaskdd** [comment]
Pushed two commits directly to this branch as a maintainer follow-up: a merge of the latest `main` (1c99f54, no conflicts) and 0539c0f, which closes a gap found while re-reviewing the v2 ownership check.

## Why the ownership check needed one more change

The post-deletion lookup went through `get_doc_by_file_basename`, and the route's `try/except` around it claimed fail-closed behavior. But that lookup has **legacy best-effort error semantics** on mongo, redis, and opensearch: a query failure is logged and swallowed into `None` (`mongo_impl.py`, `redis_impl.py`, `opensearch_impl.py` each document this). The route reads `None` as "no primary references this basename" and deletes the file — so on those backends, a transient storage read error during the check would delete a source file a live primary still uses, reproducing the exact bug this PR fixes, just on the failure path. A probe test against 8da85c5 confirmed the file is deleted (`Successfully deleted source files: paper.pdf` while `doc-primary` is alive).

## What 0539c0f changes

- The ownership check now uses `resolve_doc_source_strict`, whose base contract requires every backend to raise on failure: `SourceAbsent` → delete the file, `SourceUnique`/`SourceConflict` → preserve, raised error → the existing catch preserves (now genuinely fail-closed on all backends).
- A stale self-match (`SourceUnique` returning the just-deleted doc id from an eventually-consistent index) conservatively preserves.
- The preservation reason no longer embeds the raw exception text — `pipeline_status` is serialized into API responses (same CWE-209 hygiene as #3422); the full error goes to the log.
- Test fake grows `resolve_doc_source_strict`; the dead `get_by_id` is dropped. Three new behavioral regressions: swallowed-failure (fails on 8da85c5, passes now), unrepaired basename conflict, stale self-match.
- One sentence in `docs/FileProcessingPipeline.md` §7.3 documenting the conservative-keep behavior.

Verification: `tests/api/routes/test_document_routes_docx_archive.py` + `tests/api/test_admission_middleware.py` → 107 passed; fix-proof run with the route reverted → the new failure-path tests fail behaviorally (shared file deleted). `pre-commit` ruff/ruff-format clean.

With this, the review findings are resolved on my side.

**Comment by danielaskdd** [comment]
@codex review

File: lightrag/api/routers/document_routes.py [hunk 1]
```diff
@@ -4146,13 +4146,90 @@ async def background_delete_documents(
...
+                            #
+                            # resolve_doc_source_strict is the fail-closed
+                            # contract: a backend failure RAISES here. The
+                            # legacy get_doc_by_file_basename lookup swallows
+                            # query failures into a best-effort None on some
+                            # backends, which would read as "unreferenced" and
+                            # delete a file a live primary still uses.
+                            resolution = await rag.doc_status.resolve_doc_source_strict(
+                                result.file_path
+                            )
```
[the remark this answers is not in the document]
Comment (danielaskdd) [reply lightrag/api/routers/document_routes.py:4179]: Verified the three factual claims: the JSON `resolve_doc_source_strict` does scan the whole map on the common `SourceAbsent` outcome, `DeleteDocRequest` has no batch cap, and the scan runs under the storage lock. The severity call does not hold up, though:

1. **The complexity is not introduced by this commit.** The previous revision's `get_doc_by_file_basename` has the same O(M) miss scan on the JSON backend (`json_doc_status_impl.py:556-585` — early-return on match only; a miss scans to the end). The strict swap changed error semantics, not complexity. Relative to `main` the per-delete scan is new, but it is the cost of the fail-closed ownership check this PR exists to add.

2. **Measured, it is nowhere near "impractically slow".** At M=100k status rows, one miss scan takes ~8.3ms (strict) vs ~6.3ms (legacy) — a ~30% constant, same order. Even an N=1000 batch adds ~8s total, in a background task where each document's `adelete_by_doc_id` already performs graph purge, VDB deletes, and an input-dir walk — each individually dwarfing 8ms. The scan only runs with `delete_file=true` (default false), and every DB-backed backend (PG/Mongo/Redis/OpenSearch) resolves via an indexed query.

3. **Pre-building a batch ownership map would trade correctness for a non-bottleneck.** Deletions within one batch interact (delete the primary first, then its `dup-*` marker — an ordering the regression tests pin); a snapshot built up front reads stale state. If the JSON backend ever needs this to be faster, the right fix is a basename index in the storage layer, not bypassing the strict contract in the route.

Recording as known-and-accepted; no code change.

**Comment by danielaskdd** [comment]
@codex review
